middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -626,16 +581,92 @@ let split_and_inline env res var binding =
           bindings = Variable.Map.add var (Binding split_binding) env.bindings
         }
     in
-    let env =
+    let env, res =
       List.fold_left
-        (fun env new_binding ->
+        (fun (env, res) new_binding ->
           let flambda_var = Variable.create "to_cmm_tmp" in
-          { env with
-            bindings = Variable.Map.add flambda_var new_binding env.bindings
-          })
-        env new_bindings
+          add_binding_to_env env res flambda_var new_binding)
+        (env, res) new_bindings
+    in
+    env, res, split_binding
+
+let bind_variable_with_decision (type a) ?extra env res var ~inline
+    ~(defining_expr : a bound_expr) ~effects_and_coeffects_of_defining_expr:effs
+    =
+  (* See comment above about effects and coeffects of 'must_inline' bindings *)
+  let effs =
+    let classification =
+      To_cmm_effects.classify_by_effects_and_coeffects effs
     in
-    will_inline_complex env res split_binding
+    match[@ocaml.warning "-4"] (inline : a inline), classification with
+    | (Must_inline_once | Must_inline_and_duplicate), Generative_immutable -> (
+      (* Even if the primitive/top of the cmm expr being bound, must be
+         duplicated (as specified by [inline]), that doesn't mean that its
+         arguments (some of which may have been inlined) are also duplicatable,
+         so we must take care of only downgrading the effects and coeffects to
+         pure, and not the placement, which must be kept. *)
```

Comment:
This comment should mention that we're potentially dropping the effects of the primitive arguments. It only happens for `Generative_immutable` effects, so the only risk is moving an allocation past a `Gc` function call, but it's better to be explicit about it.

Reply:
done